README.md
[![Build Status](https://travis-ci.com/Acornsgrow/rack-vitals.svg?token=4fvrruF3QcAS3qNYjLiE)](https://travis-ci.com/Acornsgrow/rack-vitals)
[![Code Climate](https://codeclimate.com/repos/565f493f9769c16d57000dfa/badges/faec219f166eb3cb676e/gpa.svg)](https://codeclimate.com/repos/565f493f9769c16d57000dfa/feed)
[![Test Coverage](https://codeclimate.com/repos/565f493f9769c16d57000dfa/badges/faec219f166eb3cb676e/coverage.svg)](https://codeclimate.com/repos/565f493f9769c16d57000dfa/coverage)

# Rack::Vitals

Vitals is a gem that adds rack middleware to your rack applications for
checking its health.

## Installation

Add this line to your application's Gemfile with your desired version number.
Be sure to have our gem source set correctly to [gemfury.com](https://gemfury.com)'s settings.

```ruby
gem "rack-vitals", "~> 0.1"
```

And then execute:

```bash
$ bundle install
```

## Usage

#### Add `rack-vitals` to your middleware stack.

##### Rack app

```ruby
# config.ru
require 'rack/vitals'

use Rack::Vitals
run YourApp
```

##### Lotus

You can read about rails middleware in their [guide](http://lotusrb.org/guides/actions/rack-integration/).

In `config.ru`, same as a [Rack app](#rack-app).

##### Rails

You can read about rails middleware in their [guide](http://guides.rubyonrails.org/rails_on_rack.html).

```ruby
# config/application.rb
config.middleware.use Rack::Vitals
```

#### Declare the dependencies, if any, that you'd like to be checked.

```ruby
Rack::Vitals.register_checks do
  check "name of dependency", critical: true do
    if "some logic to check dependency"
      up
    else
      down
    end
  end

  check "name of dependency" do
    if "some logic to check dependency"
      up
    elsif "some other logic"
      warn
    else
      down
    end
  end
end
```

Declare one `Rack::Vitals.register_checks` block to hold all of your declared checks.

* ##### `register_checks`

  * `register_checks` takes a block where all `check` definitions are
    declared.

You can define a new check by using the `check` method within the block.

* ##### `check`

  * `check` needs a name to identify it in the `/status` check.
    ```ruby
    check "name of check here" do
    end
    ```

  * You can pass an optional argument of `critical: true` to have the check run
    for each `/health` request. This will cause the `/health` request to fail if a
    `down` is reached.
    ```ruby
    check "name of check here", critical: true do
    end
    ```

  * `check` takes a block which defines the logic that you want to run when
    the check is processed. This block should call `up`, `warn`,
    and/or `down` based on how you'd like `/status` and `/health` to
    report the checks.
    ```ruby
    check "name", critical: true do
      # Do whatever logic you want to check
      if true
        up
      else
        down
      end
    end
    ```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then,
run `rake spec` to run the tests.

#### Releasing a new version

- Update the version number in `version.rb`.
- Create a git tag for that version.
- Push git commits and tags.
- Run `gem build rack-vitals.gemspec`.
- Push the created `.gem` file to [gemfury.com](https://gemfury.com).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/acornsgrow/rack-vitals.

This project is intended to be a safe, welcoming space for collaboration, and
contributors are expected to adhere to the [Contributor Covenant](contributor-covenant.org) code of conduct.

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).



CHANGELOG.md
# ChangeLog

The following are lists of the notable changes included with each release.
This is intended to help keep people informed about notable changes between
versions as well as provide a rough history.

#### Next Release

#### v0.5.0 - 2017-02-24
- Use ENV['PATH_INFO'] instead of request path

#### v0.4.0 - 2015-12-20
- Relax rack version dependency to >= 1.6 from ~> 1.6

#### v0.3.0 - 2015-12-15
- Fix load order issue that was preventing the gem from working in an app

#### v0.2.0 - 2015-12-9
- You can add "checks" for `/health` by using the `critical: true` flag
- You can see a report of all your checks through `/status`.


spec/spec_helper.rb
require "codeclimate-test-reporter"
CodeClimate::TestReporter.start
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'rack/vitals'


spec/rack/vitals_spec.rb
require 'spec_helper'

describe Rack::Vitals do
  let(:app) { double 'app' }
  subject { Rack::Vitals.new(app) }

  it 'has a version number' do
    expect(Rack::Vitals::VERSION).not_to be nil
  end

  describe "#initialize" do
    it "initializes the app" do
      expect(subject.instance_variable_get(:@app)).to eql app
    end
  end

  describe "#call" do
    let(:env) { double("env") }
    let(:dup_middleware) { double("duplicated middleware") }

    it "duplicates itself" do
      allow(dup_middleware).to receive(:_call).with(env)
      expect(subject).to receive(:dup).and_return(dup_middleware)
      subject.call(env)
    end

    it "executes the duplicated instance of call" do
      allow(subject).to receive(:dup).and_return(dup_middleware)
      expect(dup_middleware).to receive(:_call).with(env)
      subject.call(env)
    end
  end

  describe "#_call" do
    let(:env) { double }
    let(:request) { double "Request", path_info: "/somepath" }

    before do
      allow(Rack::Request).to receive(:new).and_return request
      allow(subject).to receive(:health_vitals_response)
    end

    it "creates a rack request" do
      allow(subject).to receive(:requested_health_path?).and_return(true)
      expect(Rack::Request).to receive(:new).with(env).and_return(request)
      subject._call env
    end

    it "checks if the requested path is for the health check" do
      expect(subject).to receive(:requested_health_path?).with(request.path_info).and_return(true)
      subject._call env
    end

    context "when the requested path is for the health check" do
      before do
        allow(subject).to receive(:requested_health_path?).and_return(true)
      end

      it "gets the health vitals response" do
        expect(subject).to receive(:health_vitals_response)
        subject._call env
      end
      
      it "returns the health vitals response" do
        valid_response = double
        allow(subject).to receive(:health_vitals_response).and_return(valid_response)
        result = subject._call env
        expect(result).to eql valid_response
      end

      it "does not call the rest of the middleware stack" do
        expect(app).not_to receive(:call)
        subject._call env
      end
    end

    context "when the requested path is not for the health check" do
      context "when the request path is for the status check" do
        before do
          allow(subject).to receive(:requested_status_path?).with(request.path_info).and_return(true)
        end

        it "checks if the requested path is for the status check" do
          expect(subject).to receive(:requested_status_path?).with(request.path_info).and_return(true)
          subject._call env
        end

        it "gets the status vitals response" do
          expect(subject).to receive(:status_vitals_response)
          subject._call env
        end

        it "returns the status vitals response" do
          valid_response = double
          allow(subject).to receive(:status_vitals_response).and_return(valid_response)
          result = subject._call env
          expect(result).to eql valid_response
        end

        it "does not call the rest of the middleware stack" do
          expect(app).not_to receive(:call)
          subject._call env
        end
      end

      context "when the request path is not for the status check" do
        before do
          allow(request).to receive(:path_info).and_return("/foo")
          allow(subject).to receive(:requested_health_path?).and_return(false)
        end

        it "passes the request to the the next middleware" do
          expect(app).to receive(:call).with env
          subject._call env
        end

        it "returns the result of the next middleware layer" do
          middleware_response = double
          allow(app).to receive(:call).and_return(middleware_response)
          result = subject._call env
          expect(result).to eql middleware_response
        end
      end
    end
  end

  describe "#requested_health_path?" do
    context "when the path is for '/health'" do
      it "returns true" do
        result = subject.requested_health_path?("/health")
        expect(result).to be_truthy
      end
    end

    context "when the path is for '/health/'" do
      it "returns true" do
        result = subject.requested_health_path?("/health/")
        expect(result).to be_truthy
      end
    end

    context "when the path is for anything else" do
      it "returns false" do
        result = subject.requested_health_path?("/fooo/")
        expect(result).to be_falsey
      end
    end
  end

  describe "#requested_status_path?" do
    context "when the path is for '/status'" do
      it "returns true" do
        result = subject.requested_status_path?("/status")
        expect(result).to be_truthy
      end
    end

    context "when the path is for '/status/'" do
      it "returns true" do
        result = subject.requested_status_path?("/status/")
        expect(result).to be_truthy
      end
    end

    context "when the path is for anything else" do
      it "returns false" do
        result = subject.requested_status_path?("/fooo/")
        expect(result).to be_falsey
      end
    end
  end

  describe ".register_checks" do
    let(:registrar) { instance_double(::Rack::Vitals::CheckRegistrar) }

    before do
      allow(Rack::Vitals).to receive(:registrar).and_return(registrar)
      allow(registrar).to receive(:register)
    end

    it "gets the registrar" do
      expect(Rack::Vitals).to receive(:registrar)
      Rack::Vitals.register_checks
    end

    it "registers the give block of checks" do
      expect(registrar).to receive(:register)
      Rack::Vitals.register_checks
    end
  end

  describe ".registrar" do
    context "when it's never been called before" do
      it "creates a new registrar" do
        vitals = Class.new ::Rack::Vitals
        expect(::Rack::Vitals::CheckRegistrar).to receive(:new)
        vitals.registrar
      end
    end

    context "when it's been called before" do
      it "returns the previously created registrar" do
        vitals = Class.new ::Rack::Vitals
        registrar = vitals.registrar
        result = vitals.registrar
        expect(result).to eql registrar
      end
    end
  end

  describe "#health_vitals_response" do
    let(:detector) { instance_double Rack::Vitals::Detector }
    let(:registrar) { instance_double ::Rack::Vitals::CheckRegistrar }

    before do
      allow(::Rack::Vitals::Detector).to receive(:new).and_return detector
      allow(detector).to receive(:critical_checks_healthy?)
      allow(::Rack::Vitals).to receive(:registrar).and_return(registrar)
    end

    it "creates a new vitals detector" do
      expect(Rack::Vitals::Detector).to receive(:new).with(registrar)
      subject.health_vitals_response
    end

    it "checks all the critical checks in the registrar" do
      expect(detector).to receive(:critical_checks_healthy?)
      subject.health_vitals_response
    end

    context "when the critical checks are healthy" do
      it "returns a healthy response" do
        allow(detector).to receive(:critical_checks_healthy?).and_return true
        result = subject.health_vitals_response
        expect(result).to match_array [200, {}, ["OK"]]
      end
    end

    context "when the critical checks are not healthy" do
      it "returns an unhealthy response" do
        allow(detector).to receive(:critical_checks_healthy?).and_return false
        result = subject.health_vitals_response
        expect(result).to match_array [503, {}, ["Service Unavailable"]]
      end
    end
  end

  describe "#status_vitals_response" do
    let(:detector) { instance_double Rack::Vitals::Detector }
    let(:response_body) { double }
    let(:registrar) { instance_double ::Rack::Vitals::CheckRegistrar }

    before do
      allow(::Rack::Vitals::Detector).to receive(:new).and_return detector
      allow(detector).to receive(:generate_status_response).and_return(response_body)
      allow(::Rack::Vitals).to receive(:registrar).and_return(registrar)
    end

    it "creates a new vitals detector" do
      expect(Rack::Vitals::Detector).to receive(:new).with(registrar)
      subject.status_vitals_response
    end

    it "generates the response body for the response" do
      expect(detector).to receive(:generate_status_response)
      subject.status_vitals_response
    end

    it "returns a status response" do
      result = subject.status_vitals_response
      expect(result).to match_array [200, {"Content-Type" => "application/json"}, [response_body]]
    end
  end
end


spec/features/vitals_spec.rb
require 'spec_helper'
require 'rack/test'

describe "rack Health middleware" do
  include Rack::Test::Methods

  def app
    Rack::Builder.app do
      use Rack::Vitals
      run lambda { |env| [123, {}, ["foo"]] }
    end
  end

  after do
    ::Rack::Vitals.instance_variable_set(:@registrar, nil)
  end

  context "when the request is made to '/health'" do
    it "responds to the request that it's healthy" do
      get "/health"
      expect(last_response.status).to eql(200)
      expect(last_response.body).to eql("OK")
    end

    context "when the app has a working critical dependency" do
      before do
        Rack::Vitals.register_checks do
          check "some dependency", critical: true do
            up
          end
        end
      end

      it "responds to the request that it's healthy" do
        get "/health"
        expect(last_response.status).to eql(200)
        expect(last_response.body).to eql("OK")
      end
    end

    context "when the app has a critical dependency in a warn state" do
      before do
        Rack::Vitals.register_checks do
          check "some dependency", critical: true do
            warn
          end
        end
      end

      it "responds to the request that it's healthy" do
        get "/health"
        expect(last_response.status).to eql(200)
        expect(last_response.body).to eql("OK")
      end
    end

    context "when the app has a broken critical dependency" do
      before do
        Rack::Vitals.register_checks do
          check "some dependency", critical: true do
            down
          end
        end
      end

      it "responds to the request that is not healthy" do
        get "/health"
        expect(last_response.status).to eql(503)
        expect(last_response.body).to eql("Service Unavailable")
      end
    end

    context "when the check doesn't define a state for the check to be in" do
      before do
        Rack::Vitals.register_checks do
          check "some dependency", critical: true do
            # No states defined
          end
        end
      end

      it "responds the the request that is not healthy" do
        get "/health"
        expect(last_response.status).to eql(503)
        expect(last_response.body).to eql("Service Unavailable")
      end
    end
  end

  context "when the request is made to '/health/'" do
    it "responds to the request that it's healthy" do
      get "/health/"
      expect(last_response.status).to eql(200)
      expect(last_response.body).to eql("OK")
    end
  end

  context "when the request is made to '/status'" do
    before do
      Rack::Vitals.register_checks do
        check "foo", critical: true do
          up
        end

        check "bar" do
          down
        end

        check "baz" do
          warn
        end
      end
    end

    it "responds with a json object" do
      expected_body = [
        {
          name: "foo",
          state: "up"
        },
        {
          name: "bar",
          state: "down"
        },
        {
          name: "baz",
          state: "warn"
        }
      ].to_json
      get "/status"
      expect(last_response.status).to eql(200)
      expect(last_response.headers).to include({ "Content-Length" => "89", "Content-Type" => "application/json" })
      expect(last_response.body).to eql(expected_body)
    end

    context "when the status request has a trailing slash" do
      it "responds with a json object" do
        expected_body = [
          {
            name: "foo",
            state: "up"
          },
          {
            name: "bar",
            state: "down"
          },
          {
            name: "baz",
            state: "warn"
          }
        ].to_json
        get "/status/"
        expect(last_response.status).to eql(200)
        expect(last_response.headers).to include({ "Content-Length" => "89", "Content-Type" => "application/json" })
        expect(last_response.body).to eql(expected_body)
      end
    end
  end

  context "when the request is made to any other path" do
    it "passes the request through the middleware to the rest of the app" do
      get "/foo"
      expect(last_response.status).to eql(123)
      expect(last_response.body).to eql("foo")
    end
  end
end


lib/rack/vitals.rb
require "rack/vitals/check"
require "rack/vitals/check_evaluator"
require "rack/vitals/check_registrar"
require "rack/vitals/check_result"
require "rack/vitals/detector"
require "rack/vitals/version"
require "rack"

module Rack
  class Vitals
    HEALTH_PATH_NAMES = ["/health", "/health/"]
    STATUS_PATH_NAMES = ["/status", "/status/"]

    def initialize(app)
      @app = app
    end

    def call(env)
      dup._call(env)
    end

    def _call(env)
      request = Rack::Request.new(env)
      if self.requested_health_path?(request.path_info)
        return self.health_vitals_response
      elsif self.requested_status_path?(request.path_info)
        return self.status_vitals_response
      else
        return @app.call(env)
      end
    end

    def requested_health_path?(request_path)
      return HEALTH_PATH_NAMES.include?(request_path)
    end

    def requested_status_path?(request_path)
      return STATUS_PATH_NAMES.include?(request_path)
    end

    def self.register_checks &block
      self.registrar.register &block
    end

    def self.registrar
      @registrar ||= ::Rack::Vitals::CheckRegistrar.new
    end

    def health_vitals_response
      detector = ::Rack::Vitals::Detector.new(::Rack::Vitals.registrar)
      if detector.critical_checks_healthy?
        return [200, {}, ["OK"]]
      else
        return [503, {}, ["Service Unavailable"]]
      end
    end

    def status_vitals_response
      detector = ::Rack::Vitals::Detector.new(::Rack::Vitals.registrar)
      response_body = detector.generate_status_response
      return [200, { "Content-Type" => "application/json" }, [response_body]]
    end
  end
end
